### PR TITLE
list flattened resource users endpoint

### PR DESCRIPTION
**Pull request by marctalbott**
Ticket: [GAWB-3926](https://broadinstitute.atlassian.net/browse/GAWB-3926)

---

**PR checklist**

- [x] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 1]
```diff
@@ -392,4 +392,14 @@ class ResourceService(private val resourceTypes: Map[ResourceTypeName, ResourceT
     _ <- IO.fromFuture(IO(fireGroupUpdateNotification(policyId)))
     } yield ()
   }
+
+  def listAllFlattenedResourceUsers(resourceId: FullyQualifiedResourceId): Future[Set[UserIdInfo]] = {
```
Comment (Qi77Qi) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: do you wanna try change this signature into IO.
Grab me if you need some help :)

**Review by Qi77Qi — APPROVED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala [hunk 2]
```diff
@@ -245,4 +250,14 @@ trait ResourceRoutes extends UserInfoDirectives with SecurityDirectives with Sam
       })
     }
   }
+
+  def getAllResourceUsers(resource: FullyQualifiedResourceId, userInfo: UserInfo): server.Route = {
+    get {
+      requireOneOfAction(resource, Set(SamResourceActions.readPolicies), userInfo.userId) {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ResourceRoutes.scala]: it doesn't actually matter but there is a `requireAction` function

**Comment by dvoet** [comment]
would something like this work to list all the users in 1 ldap call:
```
  def listFlattenedResourceMembers(resourceId: FullyQualifiedResourceId): IO[Stream[WorkbenchUser]] = cs.evalOn(blockingEc)(
    IO(ldapSearchStream(peopleOu, SearchScope.ONE, Filter.createSubstringFilter(Attr.memberOf, "", Array.empty[String], resourceDn(resourceId)))(unmarshalUser))
  )
```

**Comment by marctalbott** [comment]
@dvoet it didn't work and I haven't yet figured out how to do it in a single query, but I will look at it again on Monday to see if I can get it down to one ldap call.

**Comment by dvoet** [comment]
I was able to get something on the commandline to work like `'(ismemberof=*resourceId=cclf,resourceType=managed-group,ou=resources,dc=dsde-prod,dc=broadinstitute,dc=org)'` then it is a question of figuring out the right code to make that.

**Review by dvoet — APPROVED** [review]

**Pull request merged**